src/render/src/handle.rs

```diff
@@ -1,85 +1,153 @@
 use std::marker::PhantomData;
-use std::sync::{Arc, mpsc};
-use std::{cmp, hash};
+use std::{ops, cmp, hash};
+use std::sync::mpsc;
 
-use core;
 use Backend;
 
-pub(crate) enum Garbage<B: Backend> {
-    ShaderLib(B::ShaderLib),
-    Buffer(B::Buffer),
-    Image(B::Image),
-    RenderTargetView(B::RenderTargetView),
-    DepthStencilView(B::DepthStencilView),
-    ConstantBufferView(B::ConstantBufferView),
-    ShaderResourceView(B::ShaderResourceView),
-    UnorderedAccessView(B::UnorderedAccessView),
-    Sampler(B::Sampler),
-}
-
 pub(crate) type GarbageSender<B> = mpsc::Sender<Garbage<B>>;
 pub(crate) type GarbageReceiver<B> = mpsc::Receiver<Garbage<B>>;
 pub(crate) fn garbage_channel<B: Backend>() -> (GarbageSender<B>, GarbageReceiver<B>) {
     mpsc::channel()
 }
 
-macro_rules! resource_handle {
-    ($handle:ident, $guard:ident, $resource:ident) => {
-        #[derive(Clone, Debug)]
-        struct $guard<B: Backend>(Option<B::$resource>, GarbageSender<B>);
+macro_rules! define_resources {
+    ($($name:ident: $info:path,)*) => {
+        pub enum Garbage<B: Backend> {
+            $( $name(B::$name), )*
+        }
+
+        pub mod inner {
```

Comment:
does it have to be `pub`?

Reply:
Since I'm using a type alias for the handle and not a wrapper, I think so. It also allows to call `resource()` and `info()` without any boilerplate on our side. It's the equivalent of the old [`buffer::Raw`](https://github.com/gfx-rs/gfx/blob/v0.16/src/core/src/buffer.rs#L24) for example.